write-fonts/src/graph/splitting.rs

```diff
@@ -188,6 +186,247 @@ fn split_off_ppf1(graph: &mut Graph, subtable: ObjectId, start: u16, end: u16) -
+                    graph,
+                    &data.offsets,
+                    &mut next_device_offset,
+                    &mut visited,
+                );
+            }
+        }
+
+        accumulated += accumulated_delta;
+        let largest_obj = coverage_size.max(class_def_1_size).max(class_def2_size);
+        let total = accumulated + coverage_size + class_def_1_size + class_def2_size
+            // largest obj packs last and can overflow
+            - largest_obj;
+
+        if total > MAX_TABLE_SIZE {
+            split_points.push(idx);
+            // split does not include this class, so add it for the next iteration
+            accumulated = BASE_SIZE + accumulated_delta;
+            coverage_size = 4 + estimator.increment_coverage_size(idx as _);
+            class_def_1_size = 4 + estimator.increment_class_def_size(idx as _);
+            visited.clear();
+        }
+    }
+
+    log::debug!("identified {} split points", split_points.len());
+    if split_points.is_empty() {
+        return None;
+    }
+
+    split_points.push(pp2.class1_count() as usize - 1);
+    // now we have a list of split points, and just need to do the splitting.
+    // note: harfbuzz does a thing here with a context type and an 'actuate_splits'
+    // method.
+
+    let mut new_subtables = Vec::new();
+    let mut prev_subtable_start = 0;
+    let mut next_device_offset = 3; // after coverage & two class defs
+    for subtable_start in split_points {
+        let subtable_end = subtable_start - 1;
+        let (new_subtable, offsets_used) = split_off_ppf2(
+            graph,
+            subtable,
+            prev_subtable_start,
+            subtable_end,
+            next_device_offset,
+        );
+        prev_subtable_start = subtable_start;
+        next_device_offset += offsets_used;
+        new_subtables.push(graph.add_object(new_subtable));
+    }
+
+    Some(new_subtables)
+}
+
+// returns the new table, + the number of non-null device offsets encountered.
+fn split_off_ppf2(
+    graph: &mut Graph,
+    subtable: ObjectId,
+    start: usize,
+    end: usize,
+    first_device_idx: usize,
+) -> (TableData, usize) {
+    // we have to do this bit manually (instead of via reparsing) because of borrowk
+    let coverage = graph.objects[&subtable].offsets.first().unwrap().object;
+    let coverage = graph.objects.get(&coverage).unwrap();
+    let coverage = coverage.reparse::<rlayout::CoverageTable>().unwrap();
+    let class_def_1 = graph.objects[&subtable].offsets[1].object;
+    let class_def_1 = graph.objects.get(&class_def_1).unwrap();
+    let class_def_1 = class_def_1.reparse::<rlayout::ClassDef>().unwrap();
+
+    let n_records = (end - start) + 1;
+    log::trace!("splitting off {n_records} class1records ({start}..={end})");
+
+    let class_map = coverage
+        .iter()
+        .filter_map(|gid| {
+            let glyph_class = class_def_1.get(gid);
+            (start..=end)
```

Comment:
I'm used to seeing a range with an exclusive end in Rust unless otherwise specified. I think this (specifically `(end - start) + 1` and `(start..=end)`) might be a bit more natural if we don't subtract one in the calling code?

Reply:
really good callout chad! There was 100% an off-by-one here, and in the process of investigating that I found two other straight-up mistakes (missing or incorrect fields) in the generated subtables.